### PR TITLE
Adds stat tracking for gRPC-Web.

### PR DESCRIPTION
**Pull request by fengli79**
Adds stat tracking for gRPC-Web.
TODO: Adds fine-grained stats for gRPC status.
TODO: Adds options to disable stats for better performance.

### CODE REVIEW

File: source/common/grpc/grpc_web_filter.cc [hunk 1]
```diff
@@ -130,5 +168,47 @@ Http::FilterTrailersStatus GrpcWebFilter::encodeTrailers(Http::HeaderMap& traile
   return Http::FilterTrailersStatus::Continue;
 }
 
+void GrpcWebFilter::setupStatTracking(const Http::HeaderMap& headers) {
```
Comment (mattklein123) [inline source/common/grpc/grpc_web_filter.cc]: This code is basically copied from https://github.com/lyft/envoy/blob/master/source/common/grpc/http1_bridge_filter.cc#L102 can we share?

File: source/common/grpc/grpc_web_filter.cc [hunk 2]
```diff
@@ -130,5 +168,47 @@ Http::FilterTrailersStatus GrpcWebFilter::encodeTrailers(Http::HeaderMap& traile
...
+    if (parts.size() != 2) {
+      return;
+    }
+    grpc_service_ = parts[0];
+    grpc_method_ = parts[1];
+    do_stat_tracking_ = true;
+  }
+}
+
+void GrpcWebFilter::chargeStat(const Http::HeaderMap& headers) {
```
Comment (mattklein123) [inline source/common/grpc/grpc_web_filter.cc]: code basically copied from bridge filter and https://github.com/lyft/envoy/blob/master/source/common/grpc/common.cc#L27. Can we share?

File: source/common/grpc/grpc_web_filter.cc (hunk 1 above)
Comment (fengli79) [inline source/common/grpc/grpc_web_filter.cc]: Done.
However, a better way could be get a super class for grpc-web filter and http11 bridge filter, whenever the target backend is gRPC. Not sure if it's highly desired. Potential it can be reused for REST to gRPC too.

File: source/common/grpc/common.cc [hunk 3]
```diff
@@ -24,13 +24,63 @@ namespace Grpc {
...
-  cluster.statsScope().counter(fmt::format("grpc.{}.{}.total", grpc_service, grpc_method)).inc();
+}
+
+void Common::chargeStat(const Upstream::ClusterInfo& cluster, const std::string& grpc_service,
+                        const std::string& grpc_method, bool success) {
+  chargeStat(cluster, "grpc", grpc_service, grpc_method, success);
+}
+
+Upstream::ClusterInfoConstSharedPtr
+Common::resolveClusterInfo(Http::StreamDecoderFilterCallbacks* decoder_callbacks,
```
Comment (htuch) [inline source/common/grpc/common.cc]: This utility seems more general than gRPC, it could be used by any HTTP filter. Can we put it somewhere in `source/common/http`?

File: source/common/grpc/common.cc [hunk 4]
```diff
@@ -24,13 +24,63 @@ namespace Grpc {
...
+                                     std::string* method) {
+  if (path == nullptr || path->value().c_str() == nullptr) {
+    return false;
+  }
+  std::vector<std::string> parts = StringUtil::split(path->value().c_str(), '/');
+  if (parts.size() != 2) {
+    return false;
+  }
+  *service = parts[0];
+  *method = parts[1];
```
Comment (htuch) [inline source/common/grpc/common.cc]: Could consider returning a `unique_ptr` to avoid string copies..

File: source/common/grpc/common.cc [hunk 5]
```diff
@@ -24,13 +24,63 @@ namespace Grpc {
...
-      .counter(
-           fmt::format("grpc.{}.{}.{}", grpc_service, grpc_method, success ? "success" : "failure"))
+      .counter(fmt::format("{}.{}.{}.total", protocol, grpc_service, grpc_method))
       .inc();
-  cluster.statsScope().counter(fmt::format("grpc.{}.{}.total", grpc_service, grpc_method)).inc();
+}
+
+void Common::chargeStat(const Upstream::ClusterInfo& cluster, const std::string& grpc_service,
+                        const std::string& grpc_method, bool success) {
+  chargeStat(cluster, "grpc", grpc_service, grpc_method, success);
```
Comment (htuch) [inline source/common/grpc/common.cc]: Is this overloaded variant really useful? It probably reduces boilerplate, but I'm thinking it should be named `chargeGrpcStat` to make it more self-documenting in code that uses it or moved to an anonymous namespace in the relevant gRPC code.

File: source/common/grpc/common.cc [hunk 6]
```diff
@@ -24,13 +24,63 @@ namespace Grpc {
...
 
-void Common::chargeStat(const Upstream::ClusterInfo& cluster, const std::string& grpc_service,
-                        const std::string& grpc_method, bool success) {
+void Common::chargeStat(const Upstream::ClusterInfo& cluster, const std::string& protocol,
+                        const std::string& grpc_service, const std::string& grpc_method,
+                        const Http::HeaderEntry* grpc_status) {
+  if (!grpc_status) {
+    return;
+  }
+  uint64_t grpc_status_code;
```
Comment (htuch) [inline source/common/grpc/common.cc]: Can we add an `enum` for the gRPC status codes? I'm going to need this for the streaming `async_client`?

File: source/common/grpc/common.cc [hunk 7]
```diff
@@ -24,13 +24,63 @@ namespace Grpc {
...
+      StringUtil::atoul(grpc_status->value().c_str(), grpc_status_code) && grpc_status_code == 0;
+  chargeStat(cluster, protocol, grpc_service, grpc_method, success);
+}
+
+void Common::chargeStat(const Upstream::ClusterInfo& cluster, const std::string& protocol,
+                        const std::string& grpc_service, const std::string& grpc_method,
+                        bool success) {
+  cluster.statsScope()
+      .counter(fmt::format("{}.{}.{}.{}", protocol, grpc_service, grpc_method,
+                           success ? "success" : "failure"))
```
Comment (htuch) [inline source/common/grpc/common.cc]: Why do we bin in the stats at this coarse granularity? It seems it would be generally useful to have a stats bucket for each gRPC status code, as it's useful when diagnosing issues to know if we're being rejected due to resource exhaustion, authentication, etc.

File: source/common/grpc/grpc_web_filter.cc [hunk 8]
```diff
@@ -3,18 +3,51 @@
...
+      Http::Headers::get().ContentTypeValues.GrpcWebText,
+      Http::Headers::get().ContentTypeValues.GrpcWebTextProto};
+  return types;
+}
+
+bool GrpcWebFilter::isGrpcWebRequest(const Http::HeaderMap& headers) {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type != nullptr) {
+    for (const std::string& valid : GrpcWebFilter::GRPC_WEB_CONTENT_TYPES()) {
+      if (valid == content_type->value().c_str()) {
```
Comment (htuch) [inline source/common/grpc/grpc_web_filter.cc]: Is there no fast hash based way to do this?

File: source/common/grpc/grpc_web_filter.cc [hunk 9]
```diff
@@ -3,18 +3,51 @@
...
+}
+
 // Implements StreamDecoderFilter.
-// TODO(fengli): Implements the subtypes of gRPC-Web content-type, like +proto, etc.
+// TODO(fengli): Implements the subtypes of gRPC-Web content-type other than proto, like +json, etc.
 Http::FilterHeadersStatus GrpcWebFilter::decodeHeaders(Http::HeaderMap& headers, bool) {
   const Http::HeaderEntry* content_type = headers.ContentType();
+  if (!isGrpcWebRequest(headers)) {
+    // TODO(fengli): Return error response.
+    return Http::FilterHeadersStatus::StopIteration;
```
Comment (htuch) [inline source/common/grpc/grpc_web_filter.cc]: This is a functional change that probably should be in a distinct PR, as it's not strictly stats related.

File: source/common/grpc/grpc_web_filter.cc [hunk 10]
```diff
@@ -130,5 +169,18 @@ Http::FilterTrailersStatus GrpcWebFilter::encodeTrailers(Http::HeaderMap& traile
   return Http::FilterTrailersStatus::Continue;
 }
 
+void GrpcWebFilter::setupStatTracking(const Http::HeaderMap& headers) {
+  cluster_ = Common::resolveClusterInfo(decoder_callbacks_, cm_);
+  if (!cluster_) {
+    return;
+  }
+  do_stat_tracking_ =
+      Common::resolveServiceAndMethod(headers.Path(), &grpc_service_, &grpc_method_);
```
Comment (htuch) [inline source/common/grpc/grpc_web_filter.cc]: If this fails, surely we have bigger problems, i.e. this isn't a valid gRPC. So, it's not just not to do stats tracking, gRPC should probably abort?

File: source/common/grpc/common.cc (hunk 7 above)
Comment (fengli79) [inline source/common/grpc/common.cc]: Agreed. I keep it as before just make sure it doesn't break existing http11 bridge and leave a TODO there. Per gRPC status code stat is more useful. @mattklein123 , can we change it?

File: source/common/grpc/common.cc (hunk 5 above)
Comment (fengli79) [inline source/common/grpc/common.cc]: It's under Grpc namespace. I think that's why this method doesn't have Grpc in its name from beginning. The overloaded variant is used in other places when backend is Grpc.
https://github.com/lyft/envoy/blob/43c59df63b682c742ff586bd6ee78a648dd218ad/source/common/tracing/lightstep_tracer_impl.cc#L164
https://github.com/lyft/envoy/blob/34d308e9839d8c10e4ba5c06e323dd35ac6149c6/source/common/grpc/rpc_channel_impl.cc#L46

File: source/common/grpc/grpc_web_filter.cc (hunk 10 above)
Comment (fengli79) [inline source/common/grpc/grpc_web_filter.cc]: Yes, it should be aborted, I leave a TODO there to return error response when request is not a valid gRPC call.

File: source/common/grpc/common.cc (hunk 6 above)
Comment (fengli79) [inline source/common/grpc/common.cc]: You can use https://github.com/grpc/grpc/blob/5253c8f9a899450397a5e46e4923d01ac9a66a27/include/grpc%2B%2B/impl/codegen/status_code_enum.h
As envoy is already depends on gRPC lib.
However, if the plan is to alway use envoy HTTP2, and remove the gRPC lib deps from envoy, we need to fork that enum and keep them in sync.
Here, I don't need an enum, as the stat tracking just need an int.

File: source/common/grpc/common.cc (hunk 7 above)
Comment (mattklein123) [inline source/common/grpc/common.cc]: I would leave success/failure, as we have dashboards/alarms based on it, and in general it's useful for ops to have course stats. If you want I would add new code specific stats like we do for HTTP. That way we have both.

File: source/common/grpc/common.cc (hunk 3 above)
Comment (fengli79) [inline source/common/grpc/common.cc]: It depends on source/common/upstream, lead to cyclic deps if move it to source/common/http.
Add a Utility class in source/common/upstream instead.

File: source/common/grpc/grpc_web_filter.cc (hunk 8 above)
Comment (fengli79) [inline source/common/grpc/grpc_web_filter.cc]: Done.

File: source/common/grpc/common.cc (hunk 7 above)
Comment (fengli79) [inline source/common/grpc/common.cc]: I would leave it without change. And will add the additional per status code stat in a follow up PR.
That need more changes in RpcChannelImpl and AsyncClient.

File: source/common/grpc/common.cc (hunk 6 above)
Comment (htuch) [inline source/common/grpc/common.cc]: Envoy doesn't depend on the gRPC lib today, only protobuf FYI. Looks like @alyssawilk is adding this enum in another PR anyway.

File: source/common/grpc/common.cc (hunk 7 above)
Comment (htuch) [inline source/common/grpc/common.cc]: SGTM.

File: source/common/grpc/grpc_web_filter.cc [hunk 11]
```diff
@@ -3,18 +3,48 @@
...
+#include "spdlog/spdlog.h"
 
 namespace Envoy {
 namespace Grpc {
 
 // Bit mask denotes a trailers frame of gRPC-Web.
 const uint8_t GrpcWebFilter::GRPC_WEB_TRAILER = 0b10000000;
 
+// Supported gRPC-Web content-types.
+const std::unordered_set<std::string>& GrpcWebFilter::GRPC_WEB_CONTENT_TYPES() const {
```
Comment (fengli79) [inline source/common/grpc/grpc_web_filter.cc]: Done.

File: source/common/grpc/grpc_web_filter.cc [hunk 12]
```diff
@@ -3,18 +3,48 @@
...
+  static const std::unordered_set<std::string> types{
```
Comment (fengli79) [inline source/common/grpc/grpc_web_filter.cc]: Done.

File: source/common/grpc/grpc_web_filter.cc [hunk 13]
```diff
@@ -3,18 +3,48 @@
...
+      Http::Headers::get().ContentTypeValues.GrpcWebProto,
+      Http::Headers::get().ContentTypeValues.GrpcWebText,
+      Http::Headers::get().ContentTypeValues.GrpcWebTextProto};
+  return types;
+}
+
+bool GrpcWebFilter::isGrpcWebRequest(const Http::HeaderMap& headers) {
+  const Http::HeaderEntry* content_type = headers.ContentType();
+  if (content_type != nullptr) {
+    return GRPC_WEB_CONTENT_TYPES().count(content_type->value().c_str());
```
Comment (fengli79) [inline source/common/grpc/grpc_web_filter.cc]: Done.

File: source/common/grpc/common.h [hunk 14]
```diff
@@ -28,12 +28,47 @@ class Common {
   /**
    * Charge a success/failure stat to a cluster/service/method.
    * @param cluster supplies the target cluster.
+   * @param protocol supplies the downstream protocol in use, either gRPC or gRPC-Web.
```
Comment (fengli79) [inline source/common/grpc/common.h]: Done by enumerate the actual valid values.
I actually prefer the http1 bridge use "http1bridge" instead of "grpc", to track the downstream protocol.
It should be:
"grpc": for grpc to grpc proxying.
"grpc-web": for grpc-web to grpc proxying.
"http1bridge": for http1bridge to grpc proxying.
"http-rest": for http/rest to grpc proxying.
However, as lyft has reporting tools relies on the http1bridge stats, I didn't change it.
@mattklein123 probably consider to change in the future?

Comment (htuch) [inline source/common/grpc/common.h]: This comment isn't updated with the valid values.

File: source/common/grpc/grpc_web_filter.cc (hunk 9 above)
Comment (htuch) [inline source/common/grpc/grpc_web_filter.cc]: Is there a test for this behavior (since it seems you want to keep it in this PR)?

**Comment by htuch** [comment]
@mattklein123 This is basically ready to ship modulo some small stuff, do you want to take a pass?

File: source/common/grpc/common.h (hunk 14 above)
Comment (mattklein123) [inline source/common/grpc/common.h]: Per the other stats comments WRT naming, let's leave that for a different PR. We can definitely change things here if it makes sense but would like to think through that in an isolated change.

File: source/common/grpc/grpc_web_filter.cc (hunk 1 above)
Comment (mattklein123) [inline source/common/grpc/grpc_web_filter.cc]: Ultimately I would like to just deprecate the http11 bridge filter once we get grpc-web fully rolled out. Not sure how long that will take. If you think it's worth it to do super class stuff that's fine with me.

File: source/common/grpc/grpc_web_filter.h [hunk 15]
```diff
@@ -14,7 +17,7 @@ namespace Grpc {
  */
 class GrpcWebFilter : public Http::StreamFilter, NonCopyable {
```
Comment (mattklein123) [inline source/common/grpc/grpc_web_filter.h]: not really related to this change, but can you look at coverage for this filter in master and make sure this commit gets it up to 100% coverage. I think you need to call onDestroy() filter method in your test fixture at minimum.

File: source/common/upstream/utility.h [hunk 16]
```diff
@@ -0,0 +1,25 @@
...
+ */
+class Utility {
+public:
+  /**
+   * Resolve the cluster info.
+   * @param decoder_callbacks supplies the decoder callback of filter.
+   * @param cm supplies the cluster manager of the filter.
+   */
+  static Upstream::ClusterInfoConstSharedPtr
+  resolveClusterInfo(Http::StreamDecoderFilterCallbacks* decoder_callbacks,
```
Comment (mattklein123) [inline source/common/upstream/utility.h]: I don't really care that much but nitting, this seems like more of an HTTP filter utility than an upstream utility. @htuch can make the call.

Comment (htuch) [inline source/common/upstream/utility.h]: I think @fengli79 had issues when putting it in `source/common/http/utility.{cc,h}` due to circular dependencies..

Comment (mattklein123) [inline source/common/upstream/utility.h]: Hmm. What circular dependency? Could possibly make a new filter_utility.cc/h. NBD either way.

File: source/common/grpc/grpc_web_filter.cc (hunk 9 above)
Comment (fengli79) [inline source/common/grpc/grpc_web_filter.cc]: Done.

File: source/common/grpc/grpc_web_filter.h (hunk 15 above)
Comment (fengli79) [inline source/common/grpc/grpc_web_filter.h]: Done.

File: source/common/grpc/common.h (hunk 14 above)
Comment (fengli79) [inline source/common/grpc/common.h]: OK, I will keep http1bridge filter unchanged. And use "grpc-web" for grpc_web_filter.

File: source/common/upstream/utility.h (hunk 16 above)
Comment (fengli79) [inline source/common/upstream/utility.h]: To me, upstream is a better place to hold it. It's not http filter specific, any filter which can convert a downstream protocol to sort of upstream may use it. Downstream protocol is not necessary to be http.

Comment (mattklein123) [inline source/common/upstream/utility.h]: The function takes an Http::StreamDecoderFilterCallbacks and looks up the cluster inside an HTTP route so it is pretty HTTP specific. :)

Comment (fengli79) [inline source/common/upstream/utility.h]: Done. Moved to http/filter_utility.h/cc.
